api-server/src/server/boot/authentication.js

```diff
@@ -3,7 +3,8 @@ import { check } from 'express-validator';
 import jwt from 'jsonwebtoken';
 import passport from 'passport';
 import { isEmail } from 'validator';
-
+import axios from 'axios';
+import rateLimit from 'express-rate-limit';
```

Comment:
Can't use this because:

> This module does not share the state with other processes/servers by default.

Reply:
What if it used a MongoDB store just like the session middleware? Will it work in that case?

Reply:
Yes - probably. I also realized that we might not be collecting end-user IP addresses.

Reply:
In principle it could work without sharing state, though it could be frustrating if a user is unlucky and keeps hitting the same machine.  That's rather unlikely, though.

Reply:
Well, yes, obviously. My point was just that that doesn't require synchronisation.

If you want a max 10 requests in 15 minutes globally you can either synchronize or allow 10/n per instance.

I'm not saying this is perfect, mind, since the same user could get routed the same VM repeatedly. I doubt many users would run into this, though.

Reply:
10 requests / 15 min from a given unique individual user (maybe email / or IP as an identifier) regardless of the instance they reach.

Reply:
If we have 2 instances which each allow 5 requests per 15 minutes from a unique user then that's 10 max.  Since it's random then, without state sharing, you could start rejecting requests after 5 so it's not identical, but the max is still 10.

Obviously there are tradeoffs (you have to adjust as you increase/decrease instances), but I'm not seeing an in principle reason why this impossible.

Reply:
> but I'm not seeing an in principle reason

It's over-engineering because now you are guessing the rate based on some fancy way to get instance count (or involve a human to set some parameter etc.).

Anyways we have entered bikeshedding territory here.

The approach that we want now is:

1. Introduce rate-limiting via simple state management or some other alternative (like debouncing, etc.) 
2. The lib in this current change claims explicitly to use others instead of them.

Reply:
> I also realized that we might not be collecting end-user IP addresses.

Won't the IP be in the request object?

> The lib in this current change claims explicitly to use others instead of them.

They say that they don't share state by default ie. in-memory store

Reply:
> They say that they don't share state by default ie. in-memory store

Cool. Let's try this as is for now and make a call later if we need to switch.

Reply:
> Cool. Let's try this as is for now and make a call later if we need to switch.

I already implemented it 😅 

> We are not forwarding the end-user IP to the API

Is there any issue in forwarding the user IP if we go about with it?

Reply:
It should be possible in theory, and I remember doing it for the forum years ago. I need to go back in time and trace the correct headers from Cloudflare and forward them to the API.